### PR TITLE
steps_per_epoch required if the generator is not a Sequence in fit_generator

### PR DESCRIPTION
**Pull request by daverio**
In fit_generator.

It was required to pass 'steps_per_epoch' if the generator is not based on 'Sequence'.
Passing steps_per_epoch=None should work if the generator have the method __len__(self).

### CODE REVIEW

**Comment by Dref360** [comment]
Why would you add __len__ to a generator?
This would mean that you have an object with a `__next__` method. At this point, why not implement a Sequence?

Generators are not safe anyway, so having the 'length' is not really telling.

Also, don't use try..except. It's a code smell.

**Comment by daverio** [comment]
My generator have __len__ and __next__.... So indeed its a Sequence...
Then why not use a Sequence:
Because i want the enqueuer to be GeneratorEnqueuer and not OrderedEnqueuer.
(as keeping warning for the not safe in the fit_generator)

Indeed adding a flag to set the enqueuer would be cleaner, but I think it's a bad idea to change the binding of fit_generator (at least I would not do it if I am not the main author).

Simplest hack solution... my changes. 

I think that it should be on any exception that you get the error message. So try: except: is what should be here, even if it is smelly.

**Comment by Dref360** [comment]
Sorry, but I don't see the use case.
Could you give me an example where you want GeneratorEnqueuer and not OrderedEnqueuer.

**Comment by daverio** [comment]
Very simple, if the generator is not an infinite sequence.
OrderedEnqueuer is expecting an infinite sequence (a 'Sequence'...), and therefore does not return any StopIteration exception.

If you look at Sequence, it is a bit more that just a sequence, it also make it periodic to ensure infinity.

**Comment by Dref360** [comment]
Sequences are not infinite, they are indexable, that's why we have __len__. The OrderedEnqueuer then index the Sequence just like an array. 

Generators are infinite and they will be iterated forever.

**Comment by daverio** [comment]
I am sorry but the code does not do that:

In data_utils.py:
line: 470:
# Arguments
        sequence: A `keras.utils.data_utils.Sequence` object.
        use_multiprocessing: use multiprocessing if True, otherwise threading
        shuffle: whether to shuffle the data at the beginning of each epoch
    """
line 369:
def __iter__(self):
        """Create an infinite generator that iterate over the Sequence."""
        while True:
            for item in (self[i] for i in range(len(self))):
                yield item
This make a `keras.utils.data_utils.Sequence` infinite!

An iterator generated by iter(Sequence) is an infinite iterator on a finite periodic sequence of lenght len.

**Comment by Dref360** [comment]
This is only used when workers=0. Is that your use case?

________________________________

**Comment by daverio** [comment]
No I am using worker=1. (with workers =0, you dont use the enqueuer)

**Comment by Dref360** [comment]
Exactly but with workers>0, iter is not called on the sequence. It's only used when workers=0.

**Comment by daverio** [comment]
Yes, then we are back to using the generatorEnqueuer which call next....

If I understand it well, the orderedEnqueuer is protected by the fact, that steps_per_epoch should be <= len(sequence). 
Then what happen is you set steps_per_epoch > len(sequence) (i know you should not do it, but would make the code crash without proper error management, this should be corrected)

**Comment by Dref360** [comment]
If step_per_epoch > len(sequence), The sequence will restart to index 0. Those are 2 independent components. The only thing an Enqueuer need is a Sequence or a generator. 

I'll close this PR for the following reasons:
*  length of an infinite generator is not UX friendly
* The length is useless for `workers > 1` which is the common usage
* We would like to phase out generators in the future since they are not safe. So adding features to them is counter-productive.
* The workaround is not hard (i.e. provide steps_per_epoch)

If you have any questions on the producer-consumers scheme in Keras, feel free to pm me on Slack.
